File: LeetCode/Medium/1140._Stone_Game_II.py
# Alice and Bob continue their games with piles of stones.  There are a number of piles arranged in a row, and each pile has a positive integer number of stones piles[i].  The objective of the game is to end with the most stones.
#
# Alice and Bob take turns, with Alice starting first.  Initially, M = 1.
#
# On each player's turn, that player can take all the stones in the first X remaining piles, where 1 <= X <= 2M.  Then, we set M = max(M, X).
#
# The game continues until all the stones have been taken.
#
# Assuming Alice and Bob play optimally, return the maximum number of stones Alice can get.
#
#
#
# Example 1:
#
# Input: piles = [2,7,9,4,4]
# Output: 10
# Explanation:  If Alice takes one pile at the beginning, Bob takes two piles, then Alice takes 2 piles again. Alice can get 2 + 4 + 4 = 10 piles in total. If Alice takes two piles at the beginning, then Bob can take all three piles left. In this case, Alice get 2 + 7 = 9 piles in total. So we return 10 since it's larger.
# Example 2:
#
# Input: piles = [1,2,3,4,5,100]
# Output: 104
#
#
# Constraints:
#
# 1 <= piles.length <= 100
# 1 <= piles[i] <= 104
# Solution Dynamic Programming O(N**3) O(N**2)
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        n: int = len(piles)

        suffix_sum: list[int] = [0] * n
        suffix_sum[-1] = piles[-1]
        for idx in range(n - 2, -1, -1):
            suffix_sum[idx] = piles[idx] + suffix_sum[idx + 1]

        dp: list[list[int]] = [[0] * (n + 1) for _ in range(n)]

        for idx in range(n - 1, -1, -1):
            for possible_m in range(1, n + 1):
                # Case when we can get all remaining stones
                if idx + (2 * possible_m) >= n:
                    dp[idx][possible_m] = suffix_sum[idx]
                    continue
                for x in range(1, (2 * possible_m) + 1):
                    dp[idx][possible_m] = max(
                        dp[idx][possible_m],
                        suffix_sum[idx] - dp[idx + x][max(possible_m, x)]
                    )

        return dp[0][1]

File: LeetCode/Medium/1140._Stone_Game_II.cpp
/*
Alice and Bob continue their games with piles of stones. There are a number of piles arranged in a row, and each pile has a positive integer number of stones piles[i]. The objective of the game is to end with the most stones.

Alice and Bob take turns, with Alice starting first.

On each player's turn, that player can take all the stones in the first X remaining piles, where 1 <= X <= 2M. Then, we set M = max(M, X). Initially, M = 1.

The game continues until all the stones have been taken.

Assuming Alice and Bob play optimally, return the maximum number of stones Alice can get.



Example 1:

Input: piles = [2,7,9,4,4]

Output: 10

Explanation:

If Alice takes one pile at the beginning, Bob takes two piles, then Alice takes 2 piles again. Alice can get 2 + 4 + 4 = 10 stones in total.
If Alice takes two piles at the beginning, then Bob can take all three piles left. In this case, Alice get 2 + 7 = 9 stones in total.
So we return 10 since it's larger.

Example 2:

Input: piles = [1,2,3,4,5,100]

Output: 104



Constraints:

1 <= piles.length <= 100
1 <= piles[i] <= 104
*/
// Solution
// C++ O(N^3) O(N^2) DynamicProgramming
class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        vector<vector<int>> memo(piles.size(), vector<int>(piles.size()));
        vector<int> suffixSum = piles;
        for (int i = suffixSum.size() - 2; i >= 0; --i)
            suffixSum[i] += suffixSum[i + 1];
        return maxStones(suffixSum, 1, 0, memo);
    }

    int maxStones(vector<int>& suffixSum, int maxTillNow, int currIndex,
                  vector<vector<int>>& memo) {
        if (currIndex + 2 * maxTillNow >= suffixSum.size())
            return suffixSum[currIndex];
        if (memo[currIndex][maxTillNow] > 0) return memo[currIndex][maxTillNow];
        int res = INT_MAX;

        for (int i = 1; i <= 2 * maxTillNow; ++i) {
            res = min(res, maxStones(suffixSum, max(i, maxTillNow),
                                     currIndex + i, memo));
        }
        memo[currIndex][maxTillNow] = suffixSum[currIndex] - res;
        return memo[currIndex][maxTillNow];
    }
};